pennylane/template.py

```diff
@@ -0,0 +1,234 @@
+        r (array[float]): length ``len(wires)`` arrays of squeezing amounts for :class:`~.Squeezing` operations
+        phi_r (array[float]): length ``len(wires)`` arrays of squeezing angles for :class:`~.Squeezing` operations
+        theta_2 (array[float]): length ``len(wires)*(len(wires)-1)/2`` array of transmittivity angles for second interferometer
+        phi_2 (array[float]): length ``len(wires)*(len(wires)-1)/2`` array of phase angles for second interferometer
+        a (array[float]): length ``len(wires)`` arrays of displacement magnitudes for :class:`~.Displacement` operations
+        phi_a (array[float]): length ``len(wires)`` arrays of displacement angles for :class:`~.Displacement` operations
+        k (array[float]): length ``len(wires)`` arrays of kerr parameters for :class:`~.Kerr` operations
+        wires (Sequence[int]): wires the layer should act on
+    """
+    Interferometer(theta=theta_1, phi=phi_1, wires=wires)
+    for i, wire in enumerate(wires):
+        Squeezing(r[i], phi_r[i], wires=wire)
+
+    Interferometer(theta=theta_2, phi=phi_2, wires=wires)
+
+    for i, wire in enumerate(wires):
+        Displacement(a[i], phi_a[i], wires=wire)
+
+    for i, wire in enumerate(wires):
+        Kerr(k[i], wires=wire)
+
+
+def Interferometer(theta, phi, wires=None): #pylint: disable=too-many-branches
+    r"""pennylane.template.Interferometer(theta, phi, wires)
+    General linear interferometer.
+
+    An instance is specified by providing ``len(wires)*(len(wires)-1)/2`` many
+    transmittivity angles and the same number of phase angles. The interferometer
+    is then implemented following the the scheme described in
+    :cite:`clements2016optimal` (Fig. 1a).
+    Beamsplitters are numbered per layer from top to bottom and ``theta[n]``
+    and ``phi[n]`` are used as the parameters of the n-th beam splitter.
+
+    .. note::
+
+       As the notion of a beamsplitter :math:`T_{m,n}(\theta,\phi)` in Eq. (1)
+       of :cite:`clements2016optimal` differs from the convention used in the
+       :class:`~.Beamsplitter` class of PennyLane, the Interferometer consists
+       not only of :class:`Beamsplitter(theta[n], phi[n], wires=[i,j]) <~.Beamsplitter>` Operations, but each of these is
+       followed by a :class:`Rotation(phi[n], wires=[i]) <~.Rotation>` Operation.
+
+    Args:
+        theta (array): length ``len(wires)*(len(wires)-1)/2`` array of transmittivity angles
+        phi (array): length ``len(wires)*(len(wires)-1)/2`` array of phase angles
+        wires (Sequence[int]): wires the Interferometer should act on
+    """
+
+    #loop over layers
+    n = 0
+    for l in range(len(wires)):
+        for k, (i, j) in enumerate(zip(wires[:-1], wires[1:])):
+            #skip even or odd pairs depending on layer
+            if (l+k)%2 == 1:
+                continue
+            Beamsplitter(theta[n], phi[n], wires=[i, j])
+            Rotation(phi[n], wires=[i])
```

Comment:
@cgogolin @co9olguy One final comment! Should the rotation parameter be `phi2`, as opposed to `theta, phi1` for the BS? I recall from the CVNN paper that the phase space rotation for each beamsplitter was independently parameterized.

Reply:
I saw that in the Clements decomposition in SF you have different angles and actually an additional layer of rotation gates in the middle of the decomposed interferometer. But if I compare the Clements paper with the PL/SF docs (see the writeup of my thinking in my previous comment on this) I am forced to conclude that they should be the same. Am I overlooking something?

Reply:
@josh146 , as we agreed, if you make the implementation consistent with the expected PR into SF, I will fix the unit tests and then this is finally ready :-)